### ISSUE TITLE
`Never` is not recognized as a subtype of `P` for `P` a type parameter

### ISSUE THREAD

**Issue by jonathan-laurent** [microsoft/pyright#9056]
The code below should typecheck since `Never` is a subtype of any type. However, it does not typecheck using Pyright  1.1.378.

```py
# pyright: strict

from abc import ABC, abstractmethod
from collections.abc import Callable
from dataclasses import dataclass
from typing import Never

@dataclass(frozen=True)
class Foo[T]:
    x: Callable[[T], int]

class Parent(ABC):
    @abstractmethod
    def f(self) -> Foo[Never]:
        pass

class Child1(Parent):
    def f(self) -> Foo[int]:  # ok
        assert False

class Child2[P](Parent):
    def f(self) -> Foo[P]:  # error with pyright 1.1.378
        assert False
```

Detailed Error:

```txt
error: Method "f" overrides class "Parent" in an incompatible manner
    Return type mismatch: base method returns type "Foo[Never]", override returns type "Foo[P@Child2]"
      "Foo[P@Child2]" is not assignable to "Foo[Never]"
        Type parameter "T@Foo" is contravariant, but "P@Child2" is not a supertype of "Never"
          Type "Never" is not assignable to type "P@Child2" (reportIncompatibleMethodOverride)
```

**Comment by Azureblade3808** [comment]
There is a false negative instead of a false positive.

It's correct that `Child2.f()` is not allowed to return `Foo[P]`, but it's incorrect that `Child1.f()` is allowed to return `Foo[int]`.

**Comment by erictraut** [comment]
@Azureblade3808, I think @jonathan-laurent is correct that this is a false positive. The type parameter for `Foo` is contravariant, at least in Python 3.12. This bug is specifically related to contravariant type parameters.

Interestingly, in Python 3.13, the type parameter for `Foo` [becomes invariant](https://github.com/python/mypy/issues/17623#issuecomment-2266312738) because Python 3.13 introduces a new auto-generated `__replace__` method for dataclasses.

**Comment by Azureblade3808** [comment]
> I think @jonathan-laurent is correct that this is a false positive. The type parameter for Foo is contravariant, at least in Python 3.12. This bug is specifically related to contravariant type parameters.

Yes, I was wrong in my previous post.

**Comment by jonathan-laurent** [comment]
Thanks for the quick reply! Regarding the type parameter for Foo becoming invariant in Python 3.13 and looking at the issue you linked to, I imagine there is a good argument here for special-casing `__replace__`. Indeed, changing the inferred variance of `Foo` here would break a lot of code.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.382.
